prac_07/project_management.py

```diff
@@ -0,0 +1,183 @@
+        print(HEADER, file=out_file)
+        for project in projects:
+            print(f"{project.name}\t{project.start_date}\t{project.priority}\t"
+                  f"{project.cost_estimate}\t{project.completion_estimate}", file=out_file)
+
+
+def display(projects):
+    """Display projects"""
+    print("Incomplete projects:")
+    incomplete_projects = [project for project in projects if not project.is_complete()]
+    incomplete_projects.sort()
+    for project in incomplete_projects:
+        print(f"\t{project}")
+    print("Complete projects:")
+    complete_projects = [project for project in projects if project.is_complete()]
+    complete_projects.sort()
+    for project in complete_projects:
+        print(f"\t{project}")
+
+
+def display_listed_projects(projects):  # add number to list
+    for project in enumerate(projects):
+        print(project[0], project[1])
+
+
+def add(projects):
+    """Add project"""
+    print("Let's add a new project")
+    name = input("Name: ")
+    start_date = get_valid_date("Start date (dd/mm/yyyy): ")
+    priority = get_valid_integer("Priority: ")
+    cost_estimate = get_valid_float("Cost estimate: ")
+    completion_estimate = get_valid_float("Percent complete: ")
+    project = Project(name, start_date, priority, cost_estimate, completion_estimate)
+    projects.append(project)
+
+
+def get_valid_date(prompt):
+    """Get valid date"""
+    valid_date = False
+    while not valid_date:
+        try:
+            start_date = input(prompt)
+            start_date = datetime.datetime.strptime(start_date, "%d/%m/%Y").date()
+            valid_date = True
+        except ValueError:
+            print("Invalid date")
+    return start_date
+
+
+def get_valid_integer(prompt):
+    """Get valid integer"""
+    valid_integer = False
+    while not valid_integer:
+        try:
+            integer = int(input(prompt))
+            valid_integer = True
+        except ValueError:
+            return ""
+    return integer
+
+
+def get_valid_float(prompt):
+    """Get valid float"""
+    valid_float = False
+    while not valid_float:
+        try:
+            float_value = float(input(prompt))
+            valid_float = True
+        except ValueError:
+            return ""
+    return float_value
+
+
+def update_project(projects):
+    """Update project"""
+    display_listed_projects(projects)
+    name = int(input("Project choice: "))
+    project = get_project_by_number(projects, name)
+    if project is not None:
+        print(project[1])
+        project[1].completion_estimate = get_valid_integer("New Percentage: ") or project[1].completion_estimate
+        project[1].priority = get_valid_integer("New Priority: ") or project[1].priority
+    else:
+        print("Project not found")
+
+
+def get_project_by_number(projects, number):
+    """Get project by number"""
+    for project in enumerate(projects):
+        if project[0] == number:
+            return project
+    return None
+
```

Comment:
This is a very interesting way to check if the project is in the project list. Took me awhile to understand what it was doing. Maybe add comments to improve readability or fix variable names to be less vague.